src/Assistant.ts

```diff
@@ -0,0 +1,387 @@
+  say: SayFn;
+  setStatus: SetStatusFn;
+  setSuggestedPrompts: SetSuggestedPromptsFn;
+  setTitle: SetTitleFn;
+}
+
+export interface SetStatusFn {
+  (status: string): Promise<AssistantThreadsSetStatusResponse>;
+}
+
+export interface SetSuggestedPromptsFn {
+  (params: SetSuggestedPromptsArguments): Promise<AssistantThreadsSetSuggestedPromptsResponse>;
+}
+
+export interface SetSuggestedPromptsArguments {
+  prompts: [AssistantPrompt, ...AssistantPrompt[]];
+}
+
+interface AssistantPrompt {
+  title: string;
+  message: string;
+}
+
+export interface SetTitleFn {
+  (title: string): Promise<AssistantThreadsSetTitleResponse>;
+}
+
+/**
+ * Middleware
+ */
+export type AssistantThreadStartedMiddleware = Middleware<AssistantThreadStartedMiddlewareArgs>;
+export type AssistantThreadContextChangedMiddleware = Middleware<AssistantThreadContextChangedMiddlewareArgs>;
+export type AssistantUserMessageMiddleware = Middleware<AssistantUserMessageMiddlewareArgs>;
+
+export type AssistantMiddleware =
+  | AssistantThreadStartedMiddleware[]
+  | AssistantThreadContextChangedMiddleware[]
+  | AssistantUserMessageMiddleware[];
+
+export type AssistantMiddlewareArgs =
+  | AssistantThreadStartedMiddlewareArgs
+  | AssistantThreadContextChangedMiddlewareArgs
+  | AssistantUserMessageMiddlewareArgs;
+
+export interface AssistantThreadStartedMiddlewareArgs extends
+  Omit<SlackEventMiddlewareArgs<'assistant_thread_started'>, 'say'>, AssistantUtilityArgs {}
+export interface AssistantThreadContextChangedMiddlewareArgs extends
+  Omit<SlackEventMiddlewareArgs<'assistant_thread_context_changed'>, 'say'>, AssistantUtilityArgs {}
+export interface AssistantUserMessageMiddlewareArgs extends
+  Omit<SlackEventMiddlewareArgs, 'say'>, AssistantUtilityArgs {}
+
+export type AllAssistantMiddlewareArgs<T extends AssistantMiddlewareArgs = AssistantMiddlewareArgs> =
+T & AllMiddlewareArgs;
+
+/** Constants */
+const ASSISTANT_PAYLOAD_TYPES = new Set(['assistant_thread_started', 'assistant_thread_context_changed', 'message']);
+
+export class Assistant {
+  private threadContextStore: AssistantThreadContextStore;
+
+  /** 'assistant_thread_started' */
+  private threadStarted: AssistantThreadStartedMiddleware[];
+
+  /** 'assistant_thread_context_changed' */
+  private threadContextChanged: AssistantThreadContextChangedMiddleware[];
+
+  /** 'message' */
+  private userMessage: AssistantUserMessageMiddleware[];
+
+  public constructor(config: AssistantConfig) {
+    validate(config);
+
+    const {
+      threadContextStore = new DefaultThreadContextStore(),
+      threadStarted,
+      // When `threadContextChanged` method is not provided, fallback to
+      // AssistantContextStore's save method. If a custom store has also not
+      // been provided, the default save context-via-metadata approach is used.
+      // See DefaultThreadContextStore for details of this implementation.
+      threadContextChanged = (args) => threadContextStore.save(args),
```

Comment:
@seratch Open question about whether this is the behavior you expect. Namely:

- No custom store + no threadContextChanged = save, use default context store
- Custom store + no threadContextChanged = save, use custom context store
- Custom store + use threadContextChanged = do not save, use threadContextChanged
- No custom store + threadContextChanged = do not save, use threadContextChanged

Just want to make sure we're not doing too much and saving on their behalf when they don't expect or want that to happen.

Reply:
yes, this is correct

Reply:
>(args) => threadContextStore.save(args)

nit: Adding some debug logging inside the function may be even more developer-friendly